### PR TITLE
Fix support Split-op second input

### PR DESCRIPTION
**Pull request by NValerij**
### Description
Support for latest Split-operator versions with parameters passed as input. Only constant inputs (initializers) are supported for now (`sympy` needs placeholders to make calculations).

### Motivation and Context
In version 13 of Split operator its attribute `split` [was moved](https://onnx.ai/onnx/operators/text_diff_Split_11_13.html) to input. Right now it is just ignored if met (that’s a error).

### CODE REVIEW

**Review by xadupre — DISMISSED** [review]

**Comment by NValerij** [comment]
@xadupre, I’ve fixed pyright warnings to fix builds (that prevented me from merging).
Could you please approve it once again?

**Review by xadupre — APPROVED** [review]

**Comment by NValerij** [comment]
@xadupre, could you please merge it?
(I see only stub: «This pull request can be automatically merged by project collaborators
Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests.»)

**Comment by NValerij** [comment]
There is a failed test `test_gradient_correctness_chunk` now. I’ve investigated it a little bit:
* The test [checks](https://github.com/microsoft/onnxruntime/blob/0ff61f7b971b98b93f75d4bb7bbb81fb4553a73c/orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py#L1412-L1460) correctness of `torch.chunk` forward-backward pass.
* It fails with [added by me assertion](https://github.com/microsoft/onnxruntime/pull/14311/files#diff-b08185b97bd9eefaa587c3f46213db81d7f57a326a52b542f09339070c0f762bR1788) on using second `Split` input (if it exists).
* Also this test passed before (just ignoring second input).

My guesses are:
* `torch.chunk` uses the same split logic as [was added](https://onnx.ai/onnx/operators/text_diff_Split_13_18.html) in version 18 with the `num_outputs` attribute. So tests passed even ignoring this second input.
* Not `Split` implementation is incorrect and it hides the error (of ignoring second input).
* Logic for splitting (with no explicit input) was aligned to `torch.chunk` (to pass the test?), though it looks confusing (see discussion in [4742](https://github.com/onnx/onnx/issues/4742) in onnx).

Right now I don’t know how to fix `test_gradient_correctness_chunk`. Looks like pytorch export to onnx should be fixed to emit constant tensor for second input (`chunks` are not passed to `NeuralNetChunk.forward` as parameters) or as `num_outputs`, but I’m not sure.

**Comment by xadupre** [comment]
I looked into the unit test failing. It appears that torch==1.13 does not use onnx split anymore. This issue should disappear by upgrading torch. With older version, I would probably make ORTModule use an opset < 13 to use opset 12.

* I would add this information in the error message to help users select an older opset. This message should disappear when torch is upgraded to 1.13.
* I would add more unit tests in https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/test/python/onnxruntime_test_python_symbolic_shape_infer.py to cover your changes. Creating a simple model with one node Split should be enough.

A short term solution is to use opset 12 for this specific unit test. A long term solution would be to improve shape_inference from onnxruntime. It should be able to detect it is a constant.

```python
import onnx
import torch

for dim in [0, 1, -1]:
    for chunk in [1, 3]:

        class NeuralNetChunk(torch.nn.Module):
            def __init__(self, dim):
                super(NeuralNetChunk, self).__init__()
                self.dim = dim

            def forward(self, input):
                return input.chunk(chunk, dim=dim)


        n = NeuralNetChunk(3)

        N, D, H = 16, 17, 18
        x = torch.rand((N, D, H))
        name = f"d{dim}_{chunk}.onnx"
        torch.onnx.export(n, x, f"d{dim}_{chunk}.onnx", input_names=["X"], output_names=["Y"], opset_version=16)
        onx = onnx.load(name)
        
        print(torch.__version__, dim, chunk, "split" in str(onx).lower())
```

image

**Comment by NValerij** [comment]
@xadupre, @ytaous, sorry for delay — right now I’m busy with my main job. I’ll return here ASAP.

I’ll add tests and switch to 12 opset for pytorch below 1.13.

@xadupre, for now I have no idea how to «improve shape_inference from onnxruntime. It should be able to detect it is a constant.» — seem like I’ve already done it (`assert` for input to be a constant) and that’s crashed tests.

**Review by xadupre — DISMISSED** [review]

**Comment by ytaous** [comment]
pls check the failed UT - test_gradient_correctness_chunk

**Comment by xadupre** [comment]
Based on the input, infer_shape was wrong before doing your fix but the default behaviour was matching what was expected. You need to check if the second input is a constant: look for the name into the initialiers `self.initializers_` should give you the value of this constant.

**Comment by NValerij** [comment]
@xadupre, @ytaous, could you please explain me how to run those failed tests locally? (from _onnxruntime/orttraining/orttraining/test/python/_)

Sorry, I can’t found [instructions](https://onnxruntime.ai/docs/build/inferencing.html) for it. Just building with build.sh seem not working — build is not failed with my changes. I’ve tried to add `--enable_training` and `--enable_symbolic_shape_infer_tests` with no luck.

Full build command I’ve used: `./build.sh --enable_training --config RelWithDebInfo --build_shared_lib --parallel --enable_symbolic_shape_infer_tests --test`

**Review by xadupre — APPROVED** [review]

**Comment by baijumeswani** [comment]
To run the test locally, cd into the build directory and run

```sh
python -m pytest -sv orttraining_test_ortmodule_api.py -k "test_gradient_correctness_chunk"
```

**Comment by baijumeswani** [comment]
sorry, i should have mentioned it earlier. You need a training buid. You can pass in `--enable_training` as build flag.

also, after build is complete, and before running the test, please run: `python -m onnxruntime.training.ortmodule.torch_cpp_extensions.install` from inside the build directory.

**Comment by NValerij** [comment]
> --enable_training

@baijumeswani, thank you for clarification!
I’ve tried to add this flag, but build faild with this error:
```
root@f419a7219a20:/ort# ./build.sh --config RelWithDebInfo --build_wheel --parallel --enable_training
2023-03-02 09:13:22,126 tools_python_utils [INFO] - flatbuffers module is not installed. parse_config will not be available
2023-03-02 09:13:23,316 build [DEBUG] - Command line arguments:
  --build_dir /ort/build/Linux --config RelWithDebInfo --build_wheel --parallel --enable_training
2023-03-02 09:13:23,322 build [DEBUG] - Defaulting to running update, build [and test for native builds].
2023-03-02 09:13:23,322 build [INFO] - Build started
2023-03-02 09:13:23,322 util.run [INFO] - Running subprocess in '/ort'
…
2023-03-02 09:13:23,736 build [INFO] - Generating CMake build tree
2023-03-02 09:13:23,736 util.run [INFO] - Running subprocess in '/ort/build/Linux/RelWithDebInfo'
  /usr/bin/cmake /ort/cmake -Donnxruntime_RUN_ONNX_TESTS=OFF -Donnxruntime_GENERATE_TEST_REPORTS=ON -DPython_EXECUTABLE=/usr/bin/python3 -DPYTHON_EXECUTABLE=/usr/bin/python3 -Donnxruntime_USE_MIMALLOC=OFF -Donnxruntime_ENABLE_PYTHON=ON -Donnxruntime_BUILD_CSHARP=OFF -Donnxruntime_BUILD_JAVA=OFF -Donnxruntime_BUILD_NODEJS=OFF -Donnxruntime_BUILD_OBJC=OFF -Donnxruntime_BUILD_SHARED_LIB=OFF -Donnxruntime_BUILD_APPLE_FRAMEWORK=OFF -Donnxruntime_USE_DNNL=OFF -Donnxruntime_USE_NNAPI_BUILTIN=OFF -Donnxruntime_USE_RKNPU=OFF -Donnxruntime_USE_LLVM=OFF -Donnxruntime_ENABLE_MICROSOFT_INTERNAL=OFF -Donnxruntime_USE_VITISAI=OFF -Donnxruntime_USE_TENSORRT=OFF -Donnxruntime_SKIP_AND_PERFORM_FILTERED_TENSORRT_TESTS=ON -Donnxruntime_USE_TENSORRT_BUILTIN_PARSER=OFF -Donnxruntime_TENSORRT_PLACEHOLDER_BUILDER=OFF -Donnxruntime_USE_TVM=OFF -Donnxruntime_TVM_CUDA_RUNTIME=OFF -Donnxruntime_TVM_USE_HASH=OFF -Donnxruntime_USE_MIGRAPHX=OFF -Donnxruntime_CROSS_COMPILING=OFF -Donnxruntime_DISABLE_CONTRIB_OPS=OFF -Donnxruntime_DISABLE_ML_OPS=OFF -Donnxruntime_DISABLE_RTTI=OFF -Donnxruntime_DISABLE_EXCEPTIONS=OFF -Donnxruntime_MINIMAL_BUILD=OFF -Donnxruntime_EXTENDED_MINIMAL_BUILD=OFF -Donnxruntime_MINIMAL_BUILD_CUSTOM_OPS=OFF -Donnxruntime_REDUCED_OPS_BUILD=OFF -Donnxruntime_ENABLE_LANGUAGE_INTEROP_OPS=OFF -Donnxruntime_USE_DML=OFF -Donnxruntime_USE_WINML=OFF -Donnxruntime_BUILD_MS_EXPERIMENTAL_OPS=OFF -Donnxruntime_USE_TELEMETRY=OFF -Donnxruntime_ENABLE_LTO=OFF -Donnxruntime_USE_ACL=OFF -Donnxruntime_USE_ACL_1902=OFF -Donnxruntime_USE_ACL_1905=OFF -Donnxruntime_USE_ACL_1908=OFF -Donnxruntime_USE_ACL_2002=OFF -Donnxruntime_USE_ARMNN=OFF -Donnxruntime_ARMNN_RELU_USE_CPU=ON -Donnxruntime_ARMNN_BN_USE_CPU=ON -Donnxruntime_ENABLE_NVTX_PROFILE=OFF -Donnxruntime_ENABLE_TRAINING=ON -Donnxruntime_ENABLE_TRAINING_OPS=OFF -Donnxruntime_ENABLE_TRAINING_APIS=OFF -Donnxruntime_ENABLE_CPU_FP16_OPS=ON -Donnxruntime_USE_NCCL=ON -Donnxruntime_BUILD_BENCHMARKS=OFF -Donnxruntime_USE_ROCM=OFF -DOnnxruntime_GCOV_COVERAGE=OFF -Donnxruntime_USE_MPI=ON -Donnxruntime_ENABLE_MEMORY_PROFILE=OFF -Donnxruntime_ENABLE_CUDA_LINE_NUMBER_INFO=OFF -Donnxruntime_BUILD_WEBASSEMBLY=OFF -Donnxruntime_BUILD_WEBASSEMBLY_STATIC_LIB=OFF -Donnxruntime_ENABLE_WEBASSEMBLY_EXCEPTION_CATCHING=ON -Donnxruntime_ENABLE_WEBASSEMBLY_API_EXCEPTION_CATCHING=OFF -Donnxruntime_ENABLE_WEBASSEMBLY_EXCEPTION_THROWING=ON -Donnxruntime_ENABLE_WEBASSEMBLY_THREADS=OFF -Donnxruntime_ENABLE_WEBASSEMBLY_DEBUG_INFO=OFF -Donnxruntime_ENABLE_WEBASSEMBLY_PROFILING=OFF -Donnxruntime_ENABLE_EAGER_MODE=OFF -Donnxruntime_ENABLE_LAZY_TENSOR=OFF -Donnxruntime_ENABLE_EXTERNAL_CUSTOM_OP_SCHEMAS=OFF -Donnxruntime_ENABLE_CUDA_PROFILING=OFF -Donnxruntime_ENABLE_ROCM_PROFILING=OFF -Donnxruntime_USE_XNNPACK=OFF -Donnxruntime_USE_CANN=OFF -DCMAKE_TLS_VERIFY=ON -DFETCHCONTENT_QUIET=OFF -Donnxruntime_PYBIND_EXPORT_OPSCHEMA=OFF -Donnxruntime_ENABLE_MEMLEAK_CHECKER=OFF -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_PREFIX_PATH=/ort/build/Linux/RelWithDebInfo/installed
Building ONNX Runtime for x86_64
…
[ 60%] Built target onnxruntime_training
[ 60%] Building CXX object CMakeFiles/test_execution_provider.dir/ort/onnxruntime/core/providers/shared_library/provider_bridge_provider.cc.o
[ 60%] Building CXX object CMakeFiles/onnxruntime_optimizer.dir/ort/onnxruntime/core/optimizer/transformer_memcpy.cc.o
[ 60%] Building CXX object CMakeFiles/onnxruntime_optimizer.dir/ort/onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc.o
In file included from /ort/onnxruntime/core/providers/shared_library/provider_bridge_provider.cc:53:
/ort/orttraining/orttraining/core/framework/torch/refcount_tracker.h: In static member function ‘static onnxruntime::language_interop_ops::torch::RefCountTracker& onnxruntime::language_interop_ops::torch::RefCountTracker::GetInstance()’:
/ort/orttraining/orttraining/core/framework/torch/refcount_tracker.h:31:70: error: ‘struct onnxruntime::ProviderHost’ has no member named ‘GetRefCountTrackerInstance’
   31 |   static RefCountTracker& GetInstance() { return Provider_GetHost()->GetRefCountTrackerInstance(); }
      |                                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~
gmake[2]: *** [CMakeFiles/test_execution_provider.dir/build.make:76: CMakeFiles/test_execution_provider.dir/ort/onnxruntime/core/providers/shared_library/provider_bridge_provider.cc.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:2441: CMakeFiles/test_execution_provider.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
…
subprocess.CalledProcessError: Command '['/usr/bin/cmake', '--build', '/ort/build/Linux/RelWithDebInfo', '--config', 'RelWithDebInfo', '--', '-j8']' returned non-zero exit status 2.
```

Hmm, can `python3 -m pip install torch==1.13.1` be a reason for it?

**Comment by tianleiwu** [comment]
Close this since another PR is merged: https://github.com/microsoft/onnxruntime/pull/19756

**Pull request closed without merging**
